cpp/src/randomforest/randomforest.cu

```diff
@@ -340,14 +393,211 @@ RF_metrics rfClassifier<T>::cross_validate(const cumlHandle& user_handle, const
+
+		//Cleanup
+		rows_temp_storage->release(stream);
+		selected_rows.release(stream);
+		sorted_selected_rows.release(stream);
+		delete rows_temp_storage;
+	}
+}
+
+/**
+ * @brief Predict target feature for input data; regression for single feature supported.
+ * @tparam T: data type for input data (float or double).
+ * @param[in] user_handle: cumlHandle (currently unused; API placeholder)
+ * @param[in] input: test data (n_rows samples, n_cols features) in row major format. CPU pointer.
+ * @param[in] n_rows: number of  data samples.
+ * @param[in] n_cols: number of features (excluding target feature).
+ * @param[in, out] predictions: n_rows predicted labels. CPU pointer, user allocated.
+ * @param[in] verbose: flag for debugging purposes.
+ */
+template<typename T>
+void rfRegressor<T>::predict(const cumlHandle& user_handle, const T * input, int n_rows, int n_cols, T * predictions, bool verbose) const {
+
+	ASSERT(trees, "Cannot predict! No trees in the forest.");
+	ASSERT((n_rows > 0), "Invalid n_rows %d", n_rows);
+	ASSERT((n_cols > 0), "Invalid n_cols %d", n_cols);
+	ASSERT(predictions != nullptr, "Error! User has not allocated memory for predictions.");
+
+	int row_size = n_cols;
+
+	for (int row_id = 0; row_id < n_rows; row_id++) {
+
+		if (verbose) {
+			std::cout << "\n\n";
+			std::cout << "Predict for sample: ";
+			for (int i = 0; i < n_cols; i++) std::cout << input[row_id*row_size + i] << ", ";
+			std::cout << std::endl;
+		}
+
+		T sum_predictions = 0;
+
+		for (int i = 0; i < this->rf_params.n_trees; i++) {
+			//Return prediction for one sample.
+			if (verbose) {
+				std::cout << "Printing tree " << i << std::endl;
+				trees[i].print();
+			}
+			T prediction;
+			trees[i].predict(user_handle, &input[row_id * row_size], 1, n_cols, &prediction, verbose);
+			sum_predictions += prediction;
+		}
+		// Random forest's prediction is the arithmetic mean of all its decision tree predictions.
+		predictions[row_id] = sum_predictions / this->rf_params.n_trees;
+	}
+}
+
+/**
+ * @brief Predict target feature for input data and validate against ref_labels.
+ * @tparam T: data type for input data (float or double).
+ * @param[in] user_handle: cumlHandle (currently unused; API placeholder)
+ * @param[in] input: test data (n_rows samples, n_cols features) in row major format. CPU pointer.
+ * @param[in] ref_labels: label values for cross validation (n_rows elements); CPU pointer.
+ * @param[in] n_rows: number of  data samples.
+ * @param[in] n_cols: number of features (excluding target feature).
+ * @param[in, out] predictions: n_rows predicted labels. CPU pointer, user allocated.
+ * @param[in] verbose: flag for debugging purposes.
+ */
+template<typename T>
+RF_metrics rfRegressor<T>::cross_validate(const cumlHandle& user_handle, const T * input, const T * ref_labels, int n_rows, int n_cols, T * predictions, bool verbose) const {
+
+	predict(user_handle, input, n_rows, n_cols, predictions, verbose);
+
+	double abs_difference_sum = 0;
```

Comment:
It would be nice if, instead of doing one-off cross validation for each of our algorithms, we were pushing as much into reusable primitives as possible so that all the algorithms can gain the benefit. 

If predictions were pushed to / kept on device, this would also be very trivial operations to do with our primitives (see r2_score and rand_score for reference) and would also bring the benefit that element-wise operations can be done with a high amount of parallelism, rather than looping over each row in serial.

Reply:
Agreed! I'll augment the scores.h file with the various metrics our rf cross_validate methods compute.

Reply:
@myrtw  has pushed the relevant metrics in scores.h file. Thanks